### PR TITLE
Add support for connecting Windows Slave via SSH

### PR DESCRIPTION
**Pull request by zackliu**
- Support SSH when using Windows Slave

- Change launch method of built-in image from JNLP to SSH

### CODE REVIEW

File: src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java [hunk 1]
```diff
@@ -221,6 +223,7 @@ public AzureVMAgentTemplate(
             final boolean imageReference,
             final ImageReferenceTypeClass imageReferenceTypeClass,
             final String agentLaunchMethod,
+            final Boolean preInstallSsh,
```
Comment (juniwang) [inline src/main/java/com/microsoft/azure/vmagent/AzureVMAgentTemplate.java]: why not `boolean`?

File: src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java [hunk 2]
```diff
@@ -133,18 +136,31 @@ public void launch(final SlaveComputer agentComputer, final TaskListener listene
             String initScript = agent.getInitScript();
 
             // Executing script only if script is not executed even once
+            String command;
+            if (isUnix) {
+                command = "test -e ~/.azure-agent-init";
+            } else {
+                command = "dir C:\\azure-agent-init";
```
Comment (chenkennt) [inline src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java]: Be consistent with Linux? Also use .azure-agent-init

File: src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java [hunk 3]
```diff
@@ -316,7 +343,7 @@ private int executeRemoteCommand(final Session jschSession, final String command
         try {
             // If root, modify the command to set up sudo -S
             String finalCommand = null;
-            if (executeAsRoot) {
+            if (isUnix && executeAsRoot) {
```
Comment (chenkennt) [inline src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java]: Use member isUnix to pass parameter is tricky, change it to a function parameter.

File: src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java [hunk 4]
```diff
@@ -133,18 +136,31 @@ public void launch(final SlaveComputer agentComputer, final TaskListener listene
...
                 // Grab the username/pass
                 StandardUsernamePasswordCredentials creds = AzureUtil.getCredentials(agent.getVMCredentialsId());
 
-                String command = "sh " + REMOTE_INIT_FILE_NAME;
+                if (isUnix) {
+                    command = "sh " + REMOTE_INIT_FILE_NAME;
+                } else {
+                    command = "powershell " + REMOTE_INIT_FILE_NAME_WINDOWS;
+                }
                 int exitStatus = executeRemoteCommand(session, command, logger, agent.getExecuteInitScriptAsRoot(), creds.getPassword().getPlainText());
```
Comment (chenkennt) [inline src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java]: There is a known issue that powershell.exe will always return 0 if it's running a .ps1 file. So checking exit code won't work here.

File: src/main/webapp/help-initScript.html [hunk 5]
```diff
@@ -31,6 +31,14 @@
...
       <li>
+        <b>Windows via SSH</b>
+        <p>
+          You can use this <a target="_blank" href="https://raw.githubusercontent.com/Azure/azure-devops-utils/master/powershell/Jenkins-Windows-Init-Script-SSH.ps1">sample</a>.<br/>
+          Java is <b>required</b>, others are optional.<br/>
+          We recommend to use SSH rather than JNLP. For you need less init codes and get much clearer logs.
+        </p>
+      </li>
+      <li>
         <b>Windows w/JNLP</b>
```
Comment (chenkennt) [inline src/main/webapp/help-initScript.html]: Seems Linux with JNLP doesn't really work, please also update the documentation to say it's not supported.

File: src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java (hunk 3 above)
Comment (ArieShout) [inline src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java]: We have several checks on `isUnix` and a few other states. I wonder if we can create an abstract template base class with abstract methods that provide platform dependent properties / behaviors, and then create two concrete subclass for Unix and Windows respectively.

Comment (zackliu) [inline src/main/java/com/microsoft/azure/vmagent/remote/AzureVMAgentSSHLauncher.java]: I think it's too heavy. `isUnix` only related to several command line. The total behavior between Linux and Windows are the same.

File: README.md [hunk 6]
```diff
@@ -47,22 +47,26 @@ Refer to
    used during a job configuration.
 5. Select the desired region from the combo box.
 6. Select the desired VM size.
-7. Specify the Azure Storage account name. Alternatively you can leave it blank to let Jenkins create a storage
-   account by using the default name "jenkinsarmst".
-8. Specify the retention time in minutes. This defines the number of minutes Jenkins can wait before automatically
+7. Select the Storage type, either Standard_LRS or Premium_LRS. Note that some VM size only support Standard_LRS.
+8. Specify the Azure Storage account name or select a existing Storage account name. Alternatively you can leave it blank to let Jenkins create a storage
+   account by using a unique generative name.
```
Comment (chenkennt) [inline README.md]: unique generative name -> generated unique name

File: README.md [hunk 7]
```diff
@@ -47,22 +47,26 @@ Refer to
...
     when that project specifically was tied to that node. This allows a agent to be reserved for certain kinds of jobs.
-10. Specify your Image Family. Choose between two possible alternatives:
-  * use a custom user image (provide image URL and os type - note, your custom image has to be available into the same storage account in which you are going to create agent nodes);
-  * give an image reference (provide image reference by publisher, offer, sku and version).
-11. For the launch method, select SSH or JNLP.
-  * Linux agents can be launched using SSH or JNLP, though no startup script is available through JNLP.
+12. Select built-in image or use advanced image configurations.
+  * When using built-in image, you can select Windows Server 2016 or Ubuntu 16.04 with some pre-installed tools.
+  * When using advanced image configurations. Choose between two possible alternatives:
+    * use a custom user image (provide image URL and os type - note, your custom image has to be available into the same storage account in which you are going to create agent nodes);
```
Comment (chenkennt) [inline README.md]: use -> Use, give -> Give

**Pull request merged**